core/client/fs/src/main/java/alluxio/client/metrics/MetricsHeartbeatContext.java

```diff
@@ -0,0 +1,241 @@
+          // shutdown hook
+        } catch (SecurityException e) {
+          LOG.info("Not registering metrics shutdown hook due to security exception. Regular "
+              + "heartbeats will still be performed to collect metrics data, but no final "
+              + "heartbeat will be performed on JVM exit. Security exception: {}", e.toString());
+        }
+      }
+    }
+
+    String masterAddr =
+        NetworkAddressUtils.getConnectHost(NetworkAddressUtils.ServiceType.MASTER_RPC,
+            ctx.getConf());
+
+    MetricsHeartbeatContext heartbeatCtx = MASTER_METRICS_HEARTBEAT.computeIfAbsent(masterAddr,
+        (addr) -> new MetricsHeartbeatContext(masterAddr, ctx, inquireClient));
+    heartbeatCtx.addContext(appId);
+    LOG.debug("Registered metrics heartbeat with appId: {}", appId);
+  }
+
+  /**
+   * Removes an application from the metrics heartbeat.
+   *
+   * If this is the last application to be removed for a given master then it will cancel the
+   * execution of the metrics RPC for that master.
+   *
+   * If this is the last application to be removed for the whole JVM, it will shutdown the
+   * executor threadpool until another application needs to heartbeat. If the appId which is to
+   * be removed isn't found, the application will silently continue.
+   *
+   * @param masterRpcAddr the master which the appId reports to
+   * @param appId the application Id to remove
+   */
+  public static synchronized void removeHeartbeat(String masterRpcAddr, String appId) {
+    MetricsHeartbeatContext heartbeatCtx = MASTER_METRICS_HEARTBEAT.get(masterRpcAddr);
+    if (heartbeatCtx != null) {
+      heartbeatCtx.removeContext(appId);
+    }
+
+    if (MASTER_METRICS_HEARTBEAT.isEmpty()) {
+      sExecutorService.shutdown();
+      try {
+        sExecutorService.awaitTermination(5000, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        LOG.warn("Metrics heartbeat executor did not shut down in a timely manner: ", e);
+      }
+      sExecutorService = null;
+    }
+    LOG.debug("De-registered metrics heartbeat with appId: {}", appId);
+  }
+
+  /**
+   * Class that heartbeats to the metrics master before exit. The heartbeat is performed in a second
+   * thread so that we can exit early if the heartbeat is taking too long.
+   */
+  private static final class MetricsMasterSyncShutDownHook extends Thread {
+    private final Thread mLastHeartbeatThread;
+
+    /**
+     * Creates a new metrics master shutdown hook.
+     */
+    public MetricsMasterSyncShutDownHook() {
+      mLastHeartbeatThread = new Thread(() -> {
+        if (sExecutorService != null) {
+          MASTER_METRICS_HEARTBEAT.forEach((key, value) -> value.heartbeat());
+        }
+      });
+      mLastHeartbeatThread.setDaemon(true);
+    }
+
+    @Override
+    public void run() {
+      mLastHeartbeatThread.start();
+      try {
+        // Shutdown hooks should run quickly, so we limit the wait time to 500ms. It isn't the end
```

Comment:
This comment is no longer consistent with the code?

Reply:
You may have been looking at an older version of the PR. It is consistent now